src/samVisualize/rendering/navigation.ts

```diff
@@ -0,0 +1,34 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as vscode from 'vscode'
+import * as yaml from 'yaml'
+/**
+ * Maps a resource name to the line number in the template in which it is defined
+ */
+export type ResourceLineMap = {
+    [index: string]: number
+}
+/**
+ * Maps all the resources in a given CFN template to the line numbers at which they are declared
+ * @param textDocument A TextDocument containing a CFN template from which we want to extract a ResourceLineMap
+ * @returns A ResourceLineMap containing all the resources in the given template
+ */
+export function generateResourceLineMap(cfnTemplate: vscode.TextDocument): ResourceLineMap {
+    const lineMap: ResourceLineMap = {}
+    const contents = yaml.parseDocument(cfnTemplate.getText()).contents
```

Comment:
`parseDocument` can report syntax errors with the YAML, should we report these errors back to the user?

Reply:
If there are syntax errors with the YAML, those will be caught before `generateResourceLineMap` is called. This `parseDocument` call is only to retrieve the positions of each resource key.

Reply:
If the YAML is already parsed then it would be better to pass that directly into this function, though I would need to see the rest of the code to know if that makes sense or not.

Reply:
They're parsed differently. The process of creating a `GraphObject` that the rendering code can use is done with `cfn-yaml` and the template anatomy checks are done manually. `cfn-yaml` will convert the yaml string to a js object, handling the tags, and then this object is DFS'd to find connections between resources. Technically it would be possible to completely rewrite the yaml string -> `GraphObject` code to use the `yaml.parseDocument` and then have the generateResourceLineMap use that parsed document. But that would require a complete overhaul that I don't think fits into the timeline. It may be an optimization that I can visit if I have time?

Reply:
I want to double check are we reading once + parsing twice for this cfn template file?

Reply:
As of now, yes. It's because the parsing to create a `GraphObject` was written using `yaml-cfn` which produces a JS object equivalent, and from there connections are found and a `GraphObject` is constructed. I found that finding line numbers of resources accurately needed a proper yaml parse (initially tried regular expressions) way after we had the `GraphObject` creation code well written. I'm thinking I'll have to rewrite the `GraphObject` code to receive a parsed yaml document to create the `GraphObject`, and then the `generateResourceLineMap` can reuse that parsed document. That's a pretty big overhaul though, and I'm worried I won't have time for it.

Reply:
I think I understand. The graph is being produced by a library that consumes YAML and outputs a format specifically for CFN. Here we are parsing the raw YAML to get the line numbers directly. My only concern is that this function (at least in this PR) does not have this context. We need to be clear that this function accepts an already parsed YAML file that just happens to be in a text document. In later PRs, it may be wise to enforce this concept through a type.

Reply:
So I've actually had to change this function to support highlighting of a clicked resource as opposed to just navigating to it. Since decorating an editor requires a range, and that's what the yaml parser outputs anyway, we drop the use of line numbers completely. Therefore, there is no need to have a `TextDocument` as input. I've also rewritten and verified the tests, but they'll be a bit hard to add manually. I believe my current suite is sufficient though, it contains both YAML and JSON templates. Now this function takes a raw YAML string as input to fetch start and end positions for each resource.